lib/connection.js

```diff
@@ -563,7 +565,7 @@ Connection.prototype.retrieve = function(type, ids, options, callback) {
     _.map(ids, function(id) {
       if (!id) { return Promise.reject(new Error('Invalid record ID. Specify valid record ID value')).thenCall(callback); }
       var url = [ self._baseUrl(), "sobjects", type, id ].join('/');
-      return self.request(url);
+      return self.request(url, options);
```

Comment:
Just do it by 
```
this.request({ method: 'GET', url: url, headers: options.headers });
```
You don't have to extend `request()`, just change `retrieve()` to pass the headers info in options - which is already done in `create()` or `upsert()`.